### ISSUE TITLE
Adding random number generation API

### ISSUE THREAD

**Issue by jakirkham** [data-apis/array-api#431]
In conjunction with the Array API, often need some form of random number generation as well. Since this would generate Arrays and be used with Arrays the user has, there is some benefit from downstream users perspective if there is a standard API for handling random number generation.

For example in scikit-learn, the following discussion ( https://github.com/scikit-learn/scikit-learn/issues/22352#issuecomment-1125663465 ) may shed some light. A couple points that came up were whether to have a state or stateless API and NumPy's old and new APIs.

**Comment by leofang** [comment]
> A couple points that came up were whether to have a state or stateless API and NumPy's old and new APIs.

I'd like to quote @rgommers from there (https://github.com/scikit-learn/scikit-learn/issues/22352#issuecomment-1126243581) because I feel it clarified a lot:
> It's not a stateless PRNG (there's no such thing)

**Comment by jakirkham** [comment]
Yeah was just summarizing the discussion up to that point.

Stateless isn't really the right term for this I think. A more correct description might be functional as many functional languages use a similar strategy. IOW something like this

```
state0 = seed(42)
state1, n = randint(state0)
state2, m = randint(state1)
```

IOW the user is responsible for tracking the state in this model and threading it through subsequent random number generation function calls.

The other model is object oriented like NumPy's where state lives in some object that gets mutated each time a random number is generated.

**Comment by rgommers** [comment]
We had a previous discussion on PRNG APIs, and IIRC no one was attached to the legacy `numpy.random` API (which most libraries currently have as their default - although TensorFlow is quite different: [tf.random.Generator](https://www.tensorflow.org/api_docs/python/tf/random/Generator), also class-based). 

The new NumPy API seems a little more user-friendly than the JAX APIs, and offers more functionality too. And it's no less safe - I think it's just as well possible to shoot yourself in the foot with JAX: just forget to manually create a new sub-key once. This is more of a philosophical difference: there's only one way of doing things in JAX, which is more verbose but the same for serial and parallel execution - while NumPy has a more concise way for serial (its default mode), but the user must remember to use a second method (equivalent to `jax.random.split`) to get the right parallel behavior.

One advantage of the JAX API is that there's only one way to do things, while NumPy has multiple ways to deal with parallelism - so more functionality, but also harder to understand or standardize. That may be partly due to providing multiple PRNG algorithms though, while JAX provides only the Threefry algorithm (not entirely true, there's a second experimental one - XLA Random Bit Generator). PyTorch for example provides Philox (which JAX also could have used) _and_ MT19937. Now a standard doesn't have to deal with exact reproducibility across libraries, but it should allow a design that allows libraries to choose their own algorithms. I think this works with either JAX's `key` (which contains an algorithm selector), or NumPy's class-based API. NumPy's `SeedSequence` is like the seeding part of JAX's `key`, and should work with any algorithm too. NumPy's `.jumped` cannot be supported by all algorithms AFAIK.

The main issues for standardization I see are:
-  (a) JAX cannot adopt any class-based API,
- (b) NumPy probably isn't going to be very receptive to yet another API, and 
- (c) it'd be a lot of work for other libraries like PyTorch to switch from their old API to any other API.

tl;dr this will not be easy to standardize

**Comment by jakirkham** [comment]
Thanks for the detailed write up Ralf! 🙏 

FWIW Dask is already working on adopting the new NumPy API ( https://github.com/dask/dask/pull/9038 )

It looks like CuPy already did this in 9.0.0 with PR ( https://github.com/cupy/cupy/pull/4177 ) (though Leo should feel free to correct me)

That all being said, maybe it is worth asking if a subset of NumPy's API might be easier to adopt and if so what that looks like.

Also the other important question here is what API is going to be most useful for downstream users. Started with scikit-learn as they make a lot of use of random number generation and creating some usable API with scikit-learn would be a win. Though maybe there are other downstream libraries that would make sense as well (perhaps statsmodels? others?).

**Comment by rkern** [comment]
Specifically, this code:

```python
# NumPy parallel
sseq = np.random.SeedSequence(entropy=seed)
child_seeds = sseq.spawn(4)
rngs = [np.random.default_rng(seed=s) for s in child_seeds]
```

would become:

```python
rngs = rng.spawn(4)
```

**Comment by jakirkham** [comment]
Sorry for a bit of a tangent here, but is calling `.spawn()` incrementally valid? Like

```python
rng1 = rng.spawn(1)
rng2 = rng.spawn(1)
...
rngn = rng.spawn(1)
```

This can come up when new children processes are created/destroyed on-demand (IOW autoscaling).

**Comment by rgommers** [comment]
I went searching through some older notes and found this from @alextp: _"I think the direction in TensorFlow is to follow JAX. Have functions for stateless random generation. They compose well. But they are not ergonomic. Layer on top of this will be something stateful. Once you do this, you introduce checkpoints etc, for determinism. So I would OK to add stateless API, but wouldn't be okay to add stateful one."_ That's from more than a year ago, so I don't know if it has been implemented like that, or anything changed in the meantime in TensorFlow. Maybe @edloper you can tell us?

**Comment by rkern** [comment]
I'm perfectly content with having 0 PRNG APIs in the standard (and far prefer 0 in the standard to having 2 in the standard). It seems like there is significant variety in what different communities need and want.

**Comment by shoyer** [comment]
I think it's equally straightforward to write an explicitly stateful RNG system like numpy.random.Generator using JAX. Here's a prototype:
``python
import jax


class JaxGenerator:
  def __init__(self, state):
    self.state = state

  def uniform(self):
    self.state, key = jax.random.split(self.state)
    return jax.random.uniform(key)
  
  def spawn(self, count):
    self.state, *keys = jax.random.split(self.state, count + 1)
    return [JaxGenerator(key) for key in keys]

  def __repr__(self):
    return f'{type(self).__name__}(state={self.state})'


def jax_default_rng(seed):
  return JaxGenerator(jax.random.PRNGKey(seed))


rng = jax_default_rng(0)
print(rng.uniform())  # 0.10536897
print(rng.uniform())  # 0.2787192
print(rng)  # JaxGenerator(state=[2384771982 3928867769])

rng2 = JaxGenerator(rng.state)
print(rng2)  # JaxGenerator(state=[2384771982 3928867769])

rngs = rng.spawn(2)
print(rngs)  # [JaxGenerator(state=[1777981902 3244208681]), JaxGenerator(state=[ 669635267 2816531647])]
```

From a JAX design perspective, stateful RNGs like this are not encouraged, because JAX's function transforms will break if you pass stateful objects into them. But you can still safely use this sort of API with JAX, as long as you're careful to create Generator objects _inside_ pure functions, e.g.,

```python
@jax.jit
@jax.vmap
def batched_random_uniform(seed):
  return jax_default_rng(seed).uniform()

print(batched_random_uniform(jnp.arange(5)))
# DeviceArray([0.10536897, 0.12568676, 0.4336251 , 0.47652578, 0.7844808 ],            dtype=float32)
```

Or explicitly keeping track of updated RNG state:
```python
@jax.jit
def explicit_state_random_uniform(state):
  rng = JaxGenerator(state)
  sample = rng.uniform()  # must happen *before* calculating new_state 
  new_state = rng.state
  return new_state, sample

state = jax.random.PRNGKey(0)
state2, sample = explicit_state_random_uniform(state)
print(state2, sample)
```

In fact, Haiku, which is one of the most popular neural net libraries in JAX, does something very similar with [haiku.next_rng_key](https://dm-haiku.readthedocs.io/en/latest/notebooks/basics.html#Rng-Keys-with-hk.next_rng_key).

These stateful interfaces are perhaps easier to misuse with JAX's functional transforms than pure functions, but are still a major improvement over global state.

**Comment by shoyer** [comment]
To bridge the gap between NumPy's and JAX's random number APIs, I would suggest slightly extending NumPy's API so it's easier to be explicit about state. Namely, we should add the `spawn()` method, and also support directly accessing/setting `.state` on `Generator` objects, rather than only on `BitGenerator`.

Here's my minimal wrapper of NumPy's RNG API with these extensions, to match the JAX API above:
```python
import numpy as np


class NumpyGenerator:

  def __init__(self, state):
    # TODO: avoid initializing this dummy Generator/BitGenerator?
    self._rng = np.random.default_rng()
    self.state = state

  def uniform(self):
    return self._rng.uniform()
  
  def spawn(self, count):
    entropy = self._rng.integers(2**63)  # TODO: better entropy?
    sseq = np.random.SeedSequence(entropy)
    child_seeds = sseq.spawn(count)
    return [numpy_default_rng(seed) for seed in child_seeds]

  @property
  def state(self):
    return self._rng.bit_generator.state

  @state.setter
  def state(self, value):
    self._rng.bit_generator.state = value

  def __repr__(self):
    return f'{type(self).__name__}(state={self.state})'


def numpy_default_rng(seed):
  return NumpyGenerator(np.random.default_rng(seed).bit_generator.state)


rng = numpy_default_rng(0)
print(rng.uniform())  # 0.6369616873214543
print(rng.uniform())  # 0.2697867137638703
print(rng)  # NumpyGenerator(state={'bit_generator': 'PCG64', 'state': {'state': 143609658456486183636066271097634410721, 'inc': 87136372517582989555478159403783844777}, 'has_uint32': 0, 'uinteger': 0})

rng2 = NumpyGenerator(rng.state)
print(rng2)  # NumpyGenerator(state={'bit_generator': 'PCG64', 'state': {'state': 143609658456486183636066271097634410721, 'inc': 87136372517582989555478159403783844777}, 'has_uint32': 0, 'uinteger': 0})

rngs = rng.spawn(2)
print(rngs)  # [NumpyGenerator(state={'bit_generator': 'PCG64', 'state': {'state': 164868448360684748498847325894109072011, 'inc': 241524822143570234404080558697197945801}, 'has_uint32': 0, 'uinteger': 0}), NumpyGenerator(state={'bit_generator': 'PCG64', 'state': {'state': 219573644955839246335449654370252341036, 'inc': 175512849395095609630857841553467033115}, 'has_uint32': 0, 'uinteger': 0})]
```

**Comment by jakirkham** [comment]
It's nice to see the alternative wrapping for perspective

> Namely, we should add the `spawn()` method, and also support directly accessing/setting `.state` on `Generator` objects, rather than only on `BitGenerator`.

Am curious how much work it would be to add these to NumPy?

**Comment by rkern** [comment]
Not much.

**Comment by jakevdp** [comment]
One note regarding JAX:

> JAX does seem to have a few gotchas with seed creation, it can't deal with high-entropy seeds apparently (at least in 0.2.27, released 18 Jan 2022):

JAX disables 64-bit data types by default, but you can enable this if you wish to use them:
https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#double-64bit-precision

If you enable 64-bit values, then 64-bit seeds will be fine.

**Comment by jakevdp** [comment]
> One other thing to point out: the [JAX docs comparing to NumPy](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#random-numbers) are wildly outdated/unfair, they use the non-recommended (global state) way of using the legacy API.

I'm happy to update these if you think it would be helpful. Whether the seed is global or a mutated `rng` object, I think the bulk of the discussion still holds.

**Comment by edloper** [comment]
I think the recommendation from the TF side would still be to use the stateless API.  (As mentioned above, a stateful API can be layered on top of the stateless one, if desired.)

**Comment by rkern** [comment]
We can build shim APIs across either implementation to get stateless or stateful APIs, but the way that each shim API needs to be implemented has its own costs, constraints, and tradeoffs. I would suggest exploring those implementation strategies and how well they actually support a set of use cases before relying on the mere existence proof of shim APIs as a reason to go ahead and standardize on one API instead of another. What would the shim APIs actually allow us to do?

Let's say a stateless API becomes the standard. `data = xp.uniform(key, low, high)`, etc. We can certainly build a stateful shim API on top that just uses the standard and manages the key splitting internally. What does this shim let us do? Can I write some Monte Carlo routines using that stateful shim API (completely ignorant of JAX) and still use that routine in JAX code? @shoyer suggests that has limitations. It seems like I'd be taking all the compromises necessary to make a stateless API without getting the corresponding benefits. Splittable PRNG schemes don't grow on trees, and are difficult to design robustly; the API standard would have to define an implementation as well, and I don't think I can recommend JAX's after investigating it. I can go into that in more detail here or somewhere else, if anyone would like.

Vice versa, we can definitely make a stateless API on top of the current implementation of `np.random` with just a bit of manual manipulation of `SeedSequence` inside the `xp.split()` function (`SeedSequence` can also be minorly rewritten to support more efficient pure-function splitting). Now that might allow me, a pure numpy user, to use a sampling routine that was written with JAX in mind (but otherwise restricted to the `xp` API). But making the stateful API standard just to allow an implementation to wrap a stateless API around it doesn't solve JAX's problem of not wanting to have a stateful API implementation in the first place (and you should not have to, IMO).

Again, I'm pretty happy for there to be no API standard on this subject. I would be interested in talking more on how to hand off from one style of API to the other, though. There's room in the `SeedSequence` design space to make easier.

**Comment by oleksandr-pavlyk** [comment]
> I can go into that in more detail here or somewhere else, if anyone would like.

I vote for presenting them here, please.

**Comment by rkern** [comment]
It's, uh, getting long. Not because my list of issues is long, just that I am long-winded and am including a lot of background.

**Comment by rkern** [comment]
## Apologies

This is long, and everyone has my apologies for that, but I want to make sure the background is laid out.

## Introduction

So first off, I want to say that the issues I am going to lay out don't make JAX's PRNG a bad one for all purposes. JAX has a fairly specific usage profile, and these are limitations that can be usefully lived within, if one is willing to. But these are limitations that I would hesitate to propagate up to general use amongst all Array API users, particularly in contexts where one doesn't get the benefits paid for by those limitations. I will mention a number of issues that I have with specific implementation choices that are in principle addressable. I don't really count those as determinative (they can indeed be fixed while still maintaining the style of API). But I do think they can be taken, as a whole, that imposing this style of API places some serious constraints on implementations that is probably out of scope of this standardization effort.

## Reviewing JAX's Design

To review, JAX's PRNG API is based around explicit splitting of the PRNG state and then pure functions to generate possibly-large arrays of data from the leaf keys. The `split(prng_key, num=2)` function takes `prng_key` and derives `num` new PRNG keys from it. Generally, you'll take the leftmost PRNG key and keep it around for later and then pass the other split PRNG keys down to functions (either those in `jax.random` or others that will do further `split()`s). The core PRNG algorithm for actually drawing random bits in those `jax.random` functions comes from [Parallel Random Numbers: As Easy as 1, 2, 3](http://dl.acm.org/citation.cfm?doid=2063405) (`[Random123]`). In particular, it uses the `threefry2x32` weak cryptographic block cipher as a primitive. This is a block cipher that takes a 64-bit plaintext block and encrypts it using the given 64-bit cipher key to get a 64-bit ciphertext. For this variant, all of the 64-bit blocks are broken up into 2 32-bit words (hence the name) so we can use cheap 32-bit arithmetic.

```python
cipher_text = threefry2x32(cipher_key, plain_text)
```

This primitive is a keyed bijection: given a fixed `cipher_key`, there is an invertible one-to-one mapping between `plain_text` and `cipher_text`. Changing the `cipher_key`, you select a different bijection.

For the core PRNG algorithm to draw random bits, we use a simple incrementing counter as each plaintext block. This is nice for GPUs because we can create that counter array and then have the GPU run `threefry2x32` in parallel across the whole thing. The JAX implementation of `threefry_random_bits()` which does this operation has some quirks that don't make too much of a difference for the quality of the pseudorandom numbers from a single call, but will show up a bit later when we talk about splitting.

In particular, instead of incrementing the counter in 64-bit blocks, it creates a 32-bit counter array, splits it in half to use the first half as the upper 32-bit word and the second half as the lower 32-bit word. That's a bit wasteful (you could instead just use `zeros(rem//2)` and `lax.iota(rem//2)` for `rem` requested words instead), but they are valid inputs. The two arrays which are output (being separately the upper and lower words of the `cipher_text` blocks) are then concatenated together end-to-end instead of reassembling those words. I assume this is done because it's cheaper to concatenate than to interleave again. This is fine as far as it goes; it doesn't hugely affect the quality of numbers coming out of one `jax.random.uniform()` call, for example. But it does wreck the bijection properties, which will become important when we talk about implementation choices for `split()`.

So far, so good. Despite that quirk, I have only one qualm with the core PRNG scheme for drawing bits in the `jax.random` functions: 64 bits is quite small. I consider 128-bit PRNGs to be _de rigeur_ today, particularly in contexts where we are talking about large amounts of parallelism (and the JAX API _enforces_ a large amount of parallelism). By itself, that doesn't have to be too much of an issue, and you get the benefit of a cheaper PRNG by using this smaller variant. But it does place a limit on the number of `jax.random` calls that you can make with different PRNG keys. Unless if you have a tightly controlled mechanism for allocating those PRNG keys to ensure that they are distinct, you are subject to Birthday Collisions, which happen at around the square-root of the state space size, in this case, about `2**32`. That's a lot! But you probably want to stay _well_ under that, maybe even the square root of the Birthday Bound (`2**16`, which is not a lot). Big distributed reinforcement learning runs should definitely be wary.

The mechanism that JAX uses to `split()` reuses this implementation of `threefry_random_bits()`. For the moment, let's just talk about `split(prng_key, num=2)`: it's convenient to label the two outputs as "left" and "right". In fact, let's define two operations `left()` and `right()` as follows:

```python
def left(prng_key):
    return jax.random.split(prng_key)[0]

def right(prng_key):
    return jax.random.split(prng_key)[1]
```

None of the following really depend on restricting ourselves to that, but it's handy to talk about things.

So we use the current `prng_key` to encrypt the 2 64-bit counter blocks and get out 2 64-bit ciphertexts to use as the left and right keys (modulo the quirks discussed above). This follows the general treatment in [Splittable pseudorandom number generators using cryptographic hashing](https://dl.acm.org/doi/10.1145/2578854.2503784) (`[Splittable2013]`) which uses the full-strength cryptographic Threefish block cipher (`[Random123]`'s Threefry is a weakened variant of this cipher). That is, we use the old key as the `cipher_key`, encrypt a counter block that's different whether it's the `left()` or `right()` branch, and take the resulting `cipher_text` as the next `prng_key` for that branch. So we're feeding back the output of this algorithm as an input to the next iteration. This is a standard construction for cryptographic hashes. In fact, people do use this construction as a way to make sequential PRNGs, essentially keep appending `0` blocks to a cryptographic hash and yielding the internal state of that hash as the random bits (equivalently, keep calling `left()` on its own output).

It has nice properties for cryptographic purposes, but it has some significant drawbacks for PRNGs used for scientific and statistical purposes. `left()` defines a non-invertible mapping (`right()` defines a different one). As discussed above, a block cipher defines a keyed bijection: fixing `cipher_key`, there is a one-to-one mapping between the `plain_text` and `cipher_text`. But if instead you fix `plain_text` (i.e. we use the same counter block in `left()` every time), then the mapping from `cipher_key` to `cipher_text` is many-to-one, and thus non-invertible. This is important in cryptography; if we could invert the `cipher_text` to get the `cipher_key`, then the cipher would allow an attacker to recover the `cipher_key` in a variety of scenarios. But iterations of non-invertible mappings have some unfortunate effects for scientific PRNGs where those security properties don't matter and other properties become more desirable.

## Damn Statistics

While there are few guarantees without knowing something about the structure of the mapping, you can calculate some useful statistics about the population of random non-invertible mappings. Particularly since the `left()` iteration is constructed from a (weak) block cipher, it is about as good of a random sample from that population as any. @imneme has a good pair of articles about the statistics of [non-invertible mappings](https://www.pcg-random.org/posts/too-big-to-fail.html) and also of [invertible mappings](https://www.pcg-random.org/posts/random-invertible-mapping-statistics.html). The state space of a random non-invertible mapping is not good for a scientific PRNG. You have a forest of trees that filter down to roots that are on cycles of various sizes. When a cycle completes, obviously, you have a Birthday Collision and start repeating values. The "rho length" statistic (the average number of iterations until a repeat is found) reflects that. Increasing the state space reduces the chance of these collisions predictably.

However, such state spaces are inherently biased and non-uniform, and this _doesn't_ go away as the state space increases. The states that follow states which have in-degrees greater than 1 will be overrepresented if you repeat with different initial seeds. States on small cycles will also be overrepresented. Small cycles will have big trees attached to them. A large fraction of states (on the order of half) are not reachable from any other state; you have to start with them as the initial seed to ever observe them. This is particularly an issue when you reduce the initial seeding space to just 32 bits, as is the default configuration of JAX. I will never recommend a PRNG based on non-invertible mappings for scientific use.

So far, I've talked about iterating `left(left(left(...)))`. Why do I care about this, even though we are going to be using the `right()` branch, too? Well `[Splittable2013]` also throws up its hands to say that talking about things like periods don't make sense for splittable PRNGs. After all, you are guaranteed to get a collision (by a pigeonhole argument) within _only 64_ steps down a full binary tree exploring all of the `left()` and `right()` branches (which sounds bad, but I'm not particularly concerned). But in practice, our trees are very left-heavy due to certain conventions (we tend to reserve the left-most key to pass on in the data-flow for further splitting and use the right key(s) to pass to be consumed in other functions). After all, to translate a serial PRNG code that calls multiple methods in sequence:

```python
x = rng.uniform(0, 1)
y = rng.uniform(10, 20)
z = rng.uniform(-1, 1)

# -->

key, subkey = split(key)
x = uniform(subkey, 0, 1)
key, subkey = split(key)
y = uniform(subkey, 10, 20)
key, subkey = split(key)
z = uniform(subkey, -1, 1)
```

So the iteration of `left()` is kind of important to characterize, IMO. There are things we can say about it, and maybe improve it.

So I did construct a PRNG out of repeated iteration of `left()` and fed its output to [PractRand](http://pracrand.sourceforge.net/). And we can observe that we do in fact find one of those cycles and fail right around the Birthday Bound (`2**35 bytes` corresponds to `2**32 iterations` yielding 8-byte keys).

<details>
<summary>PractRand output for <code>jax.random.split(key)[0]</code></summary>

```
❯ ./jax_keys.py
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
RNG_test using PractRand version 0.93
RNG = RNG_stdin64, seed = 0x6ae890b2
test set = expanded, folding = extra

rng=RNG_stdin64, seed=0x6ae890b2
length= 8 megabytes (2^23 bytes), time= 2.6 seconds
  no anomalies in 694 test result(s)

rng=RNG_stdin64, seed=0x6ae890b2
length= 16 megabytes (2^24 bytes), time= 7.8 seconds
  no anomalies in 747 test result(s)

rng=RNG_stdin64, seed=0x6ae890b2
length= 32 megabytes (2^25 bytes), time= 15.6 seconds
  no anomalies in 796 test result(s)

rng=RNG_stdin64, seed=0x6ae890b2
length= 64 megabytes (2^26 bytes), time= 28.1 seconds
  no anomalies in 843 test result(s)

rng=RNG_stdin64, seed=0x6ae890b2
length= 128 megabytes (2^27 bytes), time= 50.3 seconds
  Test Name                         Raw       Processed     Evaluation
  [Low4/32]Gap-16:B                 R=  -4.7  p =1-5.0e-4   unusual          
  ...and 890 test result(s) without anomalies

rng=RNG_stdin64, seed=0x6ae890b2
length= 256 megabytes (2^28 bytes), time= 92.0 seconds
  no anomalies in 938 test result(s)

rng=RNG_stdin64, seed=0x6ae890b2
length= 512 megabytes (2^29 bytes), time= 171 seconds
  no anomalies in 985 test result(s)

rng=RNG_stdin64, seed=0x6ae890b2
length= 1 gigabyte (2^30 bytes), time= 326 seconds
  no anomalies in 1039 test result(s)

rng=RNG_stdin64, seed=0x6ae890b2
length= 2 gigabytes (2^31 bytes), time= 628 seconds
  no anomalies in 1092 test result(s)

rng=RNG_stdin64, seed=0x6ae890b2
length= 4 gigabytes (2^32 bytes), time= 1237 seconds
  no anomalies in 1151 test result(s)

rng=RNG_stdin64, seed=0x6ae890b2
length= 8 gigabytes (2^33 bytes), time= 2439 seconds
  no anomalies in 1224 test result(s)

rng=RNG_stdin64, seed=0x6ae890b2
length= 16 gigabytes (2^34 bytes), time= 4831 seconds
  Test Name                         Raw       Processed     Evaluation
  [Low4/64]BCFN_FF(2+2,13-1,T)      R=  +9.4  p =  1.6e-4   unusual          
  ...and 1300 test result(s) without anomalies

rng=RNG_stdin64, seed=0x6ae890b2
length= 32 gigabytes (2^35 bytes), time= 9622 seconds
  Test Name                         Raw       Processed     Evaluation
  BCFN_FF(2+0,13-0,T)               R= +41.3  p =  1.3e-21    FAIL !!        
  BCFN_FF(2+1,13-0,T)               R= +33.9  p =  1.1e-17    FAIL !         
  BCFN_FF(2+2,13-0,T)               R= +44.0  p =  4.4e-23    FAIL !!        
  BCFN_FF(2+3,13-0,T)               R= +48.1  p =  3.1e-25    FAIL !!        
  BCFN_FF(2+4,13-0,T)               R= +44.8  p =  1.6e-23    FAIL !!        
  BCFN_FF(2+5,13-0,T)               R= +48.1  p =  3.0e-25    FAIL !!        
  BCFN_FF(2+6,13-0,T)               R= +40.9  p =  2.2e-21    FAIL !!        
  BCFN_FF(2+7,13-1,T)               R= +29.0  p =  4.5e-15    FAIL           
  BCFN_FF(2+8,13-1,T)               R= +30.6  p =  6.1e-16    FAIL           
  DC6-9x1Bytes-1                    R=+230.8  p =  4.5e-155   FAIL !!!!!     

[RTK: snipping 630 more failures; this is definitely a cycle]

  ...and 717 test result(s) without anomalies
```

</details>

As mentioned in the article, increasing the state size does help with this. You can hold off those collisions more and more with larger state sizes. You can hide this flaw with larger state spaces, as done in the original `[Splittable2013]` paper, which uses a full-strength 256-bit block cipher.

Just to be sure that this wasn't trivially fixable by fixing the aforementioned quirk, I retried this a version that did the `threefry2x32` manually with a fixed plaintext of `0`. This also failed at `2**36 bytes` (about the same, essentially), so confirming that the general scheme of using the old `prng_key` as the `cipher_key` is still a non-invertible mapping.

## Alternatives

There are alternatives, which are to use invertible mappings to define `left()` and `right()` (and further for `num > 2`). Random invertible mappings have [much nicer expected properties](https://www.pcg-random.org/posts/random-invertible-mapping-statistics.html#random-invertible-mapping-statistics). Every state is on a cycle and has an in-degree of 1. Given a uniform distribution of root seeds, you have an equal chance of seeing any state (if you don't draw enough to hit a cycle). We have a good idea of the distribution of sizes of those cycles. Very, very roughly, in expectation, half of the states are on the biggest cycle, 1/4 are on the next one, 1/8 on the next one and so on. There still are small cycles, but the chance that your initial seed actually places you on one goes down as the size does. For 64-bit states, the chance of your root seed being on a cycle less than `2**32` (that Birthday Bound) is `1/2**(64-32) == 1/2**32`, a pretty tiny probability. Whereas for the non-invertible mapping used in JAX, your chance of hitting a repeat in `2**32` draws is about half.

How do we construct useful invertible mappings? One way is to use `SeedSequence`, [whose construction](https://www.pcg-random.org/posts/simple-portable-cpp-seed-entropy.html) is conceptually very similar to the splitting mechanism here. Instead of a cryptographic hash, it folds data into the internal pool using invertible operations such that bijections are maintained where possible. So repeated iterations of `left()` or `right()` will at least have the invertible mapping statistics. There is also an MCG in the middle there driving it, much like `SFC64` incorporates a 64-bit counter. That has the effect of creating a minimum cycle size of at least `2**28` (I think) when we repeatedly take the leftmost spawn child.

So for example, I have here a PractRand run where I use a tiny 32-bit pool size (though this is a little misleading as the MCG has to be tracked as well, so it's really a 64-bit state size). I run it like I did the JAX splitting test, where I just repeatedly fold in the `0` counter.

<details>
<summary>PractRand output for <code>seed_sequence.spawn(1)[0]</code></summary>

```
❯ ./ss_split.py -p 1
RNG_test using PractRand version 0.93
RNG = RNG_stdin64, seed = 0x64fd6e66
test set = expanded, folding = extra

rng=RNG_stdin64, seed=0x64fd6e66
length= 1 megabyte (2^20 bytes), time= 2.1 seconds
  Test Name                         Raw       Processed     Evaluation
  [Low4/32]FPF-14+6/32:cross        R=  +6.5  p =  2.8e-5   mildly suspicious
  ...and 530 test result(s) without anomalies

rng=RNG_stdin64, seed=0x64fd6e66
length= 2 megabytes (2^21 bytes), time= 6.2 seconds
  no anomalies in 585 test result(s)

rng=RNG_stdin64, seed=0x64fd6e66
length= 4 megabytes (2^22 bytes), time= 12.4 seconds
  Test Name                         Raw       Processed     Evaluation
  [Low4/64]FPF-14+6/32:cross        R=  +6.0  p =  7.7e-5   unusual          
  ...and 640 test result(s) without anomalies

rng=RNG_stdin64, seed=0x64fd6e66
length= 8 megabytes (2^23 bytes), time= 22.8 seconds
  no anomalies in 694 test result(s)

rng=RNG_stdin64, seed=0x64fd6e66
length= 16 megabytes (2^24 bytes), time= 41.6 seconds
  no anomalies in 747 test result(s)

rng=RNG_stdin64, seed=0x64fd6e66
length= 32 megabytes (2^25 bytes), time= 77.3 seconds
  no anomalies in 796 test result(s)

rng=RNG_stdin64, seed=0x64fd6e66
length= 64 megabytes (2^26 bytes), time= 147 seconds
  no anomalies in 843 test result(s)

rng=RNG_stdin64, seed=0x64fd6e66
length= 128 megabytes (2^27 bytes), time= 280 seconds
  no anomalies in 891 test result(s)

rng=RNG_stdin64, seed=0x64fd6e66
length= 256 megabytes (2^28 bytes), time= 544 seconds
  no anomalies in 938 test result(s)

rng=RNG_stdin64, seed=0x64fd6e66
length= 512 megabytes (2^29 bytes), time= 1066 seconds
  no anomalies in 985 test result(s)

rng=RNG_stdin64, seed=0x64fd6e66
length= 1 gigabyte (2^30 bytes), time= 2106 seconds
  Test Name                         Raw       Processed     Evaluation
  [Low1/64]Gap-16:B                 R=  -5.6  p =1-7.8e-5   unusual          
  ...and 1036 test result(s) without anomalies
rng=RNG_stdin64, seed=0x64fd6e66
length= 2 gigabytes (2^31 bytes), time= 4204 seconds
  no anomalies in 1092 test result(s)

rng=RNG_stdin64, seed=0x64fd6e66
length= 4 gigabytes (2^32 bytes), time= 8436 seconds
  no anomalies in 1155 test result(s)

rng=RNG_stdin64, seed=0x64fd6e66
length= 8 gigabytes (2^33 bytes), time= 16916 seconds
  no anomalies in 1224 test result(s)

rng=RNG_stdin64, seed=0x64fd6e66
length= 16 gigabytes (2^34 bytes), time= 33704 seconds
  no anomalies in 1302 test result(s)

rng=RNG_stdin64, seed=0x64fd6e66
length= 32 gigabytes (2^35 bytes), time= 67929 seconds
  no anomalies in 1358 test result(s)

rng=RNG_stdin64, seed=0x64fd6e66
length= 64 gigabytes (2^36 bytes), time= 136044 seconds
  no anomalies in 1434 test result(s)
```

</details>

And it's still going strong past the point that the JAX one failed. Now, to be fair, this definitely created a duplicate entropy pool state somewhere along the way, which would have seeded an identical copy of the `BitGenerator` if it were used. Only the 32-bit pool plays a role in the seeding of the `BitGenerator`, not the MCG. If we actually drew a lot of numbers from each `BitGenerator`, PractRand would have found that out. It's still separately important that generated seeds that go into the number-generating API calls still have an acceptable size to avoid collisions.

Currently, `SeedSequence` is written statefully. It keeps track of the number of spawned children requested of it so that repeated calls return freshly independent `SeedSequence`s. It is also somewhat wasteful in keeping an explicit representation of the path down the split tree from the root entropy. Neither of these implementation choices are necessary to the algorithm. [Here is an implementation](https://gist.github.com/rkern/9361aa15a28ae8c6dced01840209cdbb) with an idempotent `.split()` method.

Now maybe you don't want to do all that. Maybe you don't like the expense of all of those multiplications; maybe you don't like that extra MCG hanging around taking up space. You have a keyed bijection already: `threefry2x32()`. But to get a bijection, you have to reverse the roles of the old `prng_key` and the counter block. Instead of using the old `prng_key` as the `cipher_key` and the counter block as the `plain_text`, flip it.

```python
new_prng_key = threefry2x32(counter, old_prng_key)
```

Now there is a bijection between the two `prng_key`s because that's what a block cipher does. Running the repeated `left(left(left(...)))` iteration as a PRNG is still going strong long past where the original failed.

<details>
<summary>PractRand output for a ThreeFry2x32 bijection scheme</summary>

```
❯ ./jax_keys.py --variant bijective
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
RNG_test using PractRand version 0.93
RNG = RNG_stdin64, seed = 0x7a16bac8
test set = expanded, folding = extra

rng=RNG_stdin64, seed=0x7a16bac8
length= 8 megabytes (2^23 bytes), time= 2.5 seconds
  no anomalies in 694 test result(s)

rng=RNG_stdin64, seed=0x7a16bac8
length= 16 megabytes (2^24 bytes), time= 7.6 seconds
  Test Name                         Raw       Processed     Evaluation
  BCFN_FF(2+1,13-5,T)               R= +12.1  p =  6.5e-5   unusual          
  ...and 746 test result(s) without anomalies

rng=RNG_stdin64, seed=0x7a16bac8
length= 32 megabytes (2^25 bytes), time= 15.2 seconds
  no anomalies in 796 test result(s)

rng=RNG_stdin64, seed=0x7a16bac8
length= 64 megabytes (2^26 bytes), time= 27.5 seconds
  no anomalies in 843 test result(s)

rng=RNG_stdin64, seed=0x7a16bac8
length= 128 megabytes (2^27 bytes), time= 49.1 seconds
  no anomalies in 891 test result(s)

rng=RNG_stdin64, seed=0x7a16bac8
length= 256 megabytes (2^28 bytes), time= 89.2 seconds
  no anomalies in 938 test result(s)

rng=RNG_stdin64, seed=0x7a16bac8
length= 512 megabytes (2^29 bytes), time= 166 seconds
  no anomalies in 985 test result(s)

rng=RNG_stdin64, seed=0x7a16bac8
length= 1 gigabyte (2^30 bytes), time= 318 seconds
  Test Name                         Raw       Processed     Evaluation
  [Low1/16]Gap-16:B                 R=  -4.8  p =1-3.8e-4   unusual          
  ...and 1035 test result(s) without anomalies

rng=RNG_stdin64, seed=0x7a16bac8
length= 2 gigabytes (2^31 bytes), time= 616 seconds
  Test Name                         Raw       Processed     Evaluation
  [Low4/64]FPF-14+6/4:all           R=  -4.8  p =1-2.8e-4   unusual          
  ...and 1091 test result(s) without anomalies

rng=RNG_stdin64, seed=0x7a16bac8
length= 4 gigabytes (2^32 bytes), time= 1211 seconds
  no anomalies in 1155 test result(s)

rng=RNG_stdin64, seed=0x7a16bac8
length= 8 gigabytes (2^33 bytes), time= 2405 seconds
  no anomalies in 1229 test result(s)

rng=RNG_stdin64, seed=0x7a16bac8
length= 16 gigabytes (2^34 bytes), time= 4777 seconds
  no anomalies in 1303 test result(s)

rng=RNG_stdin64, seed=0x7a16bac8
length= 32 gigabytes (2^35 bytes), time= 9522 seconds
  Test Name                         Raw       Processed     Evaluation
  [Low4/64]BCFN_FF(2+1,13-0,T)      R=  +9.5  p =  1.3e-4   unusual          
  ...and 1359 test result(s) without anomalies

rng=RNG_stdin64, seed=0x7a16bac8
length= 64 gigabytes (2^36 bytes), time= 18960 seconds
  Test Name                         Raw       Processed     Evaluation
  [Low1/32]BCFN_FF(2+0,13-0,T)      R=  +8.5  p =  4.4e-4   unusual          
  [Low4/32]FPF-14+6/32:all          R=  +5.5  p =  1.3e-4   unusual          
  ...and 1429 test result(s) without anomalies

rng=RNG_stdin64, seed=0x7a16bac8
length= 128 gigabytes (2^37 bytes), time= 37487 seconds
  no anomalies in 1506 test result(s)

rng=RNG_stdin64, seed=0x7a16bac8
length= 256 gigabytes (2^38 bytes), time= 74451 seconds
  no anomalies in 1567 test result(s)
```

</details>

Now, that's neat (I think). But we don't just do iterations of `left()`. We mix up `left()` and `right()` (and more). Each of those are different bijections. There is still a chance that a `right()` call will give you an earlier-seen state. While the structure provided by bijections lets you have a better safety factor on parts of the split tree (important parts, IMO), it doesn't help much for arbitrarily mixed operations. With the statistics of random bijections, I think you could probably derive more accurate bounds, but even with `SeedSequence`, we tell you to rely on the Birthday Bound in any case. But you still get the statistical benefits of using bijections.

## Conclusion

I want to close with a quote from the conclusion of [Evaluation of Splittable Pseudo-Random Generators](http://www.hg.schaathun.net/research/Papers/hgs2015jfp.pdf):

> Conventional wisdom in computational statistics and Monte Carlo simulation indicate that one PRNG cannot suit every purpose. Different applications have different priorities, and calling for different trade-offs between sequence length, distribution, running time, and theoretical guarantees.

The JAX style of PRNG is a very _specific_ flavor of splittability and pure functionalism. There are splittable PRNGs that are not functional, and there are pure functional APIs that don't force explicit splitting (and of course, useful PRNG schemes that are neither). JAX has a specific profile of needs that make this specific flavor of PRNG very appealing for its use cases. However, JAX's profile is not universal and should not _define_ the API standard, which has a broad scope. For that matter, the profile behind the current design of `np.random` is not universal either.

I am perfectly content with this functionality not being standardized by the Array API.

**Comment by jakirkham** [comment]
It's probably also worthwhile to tie this back to how downstream users would leverage this API.

Going back to scikit-learn, many APIs take a `RandomState` object (for [example]( https://scikit-learn.org/stable/modules/generated/sklearn.cluster.AffinityPropagation.html#sklearn.cluster.AffinityPropagation )). This of course is referencing the old NumPy API, but that's not really the point (as NumPy's new API is also object-based). Instead the point is scikit-learn depends on having an object to pass state around and generate random numbers from different distributions (reusing [the example above]( https://github.com/scikit-learn/scikit-learn/blob/80598905e517759b4696c74ecc35c6e2eb508cff/sklearn/cluster/_affinity_propagation.py#L174 )). Though this isn't unique to scikit-learn, statsmodels also does this (for [example]( https://www.statsmodels.org/devel/generated/statsmodels.distributions.copula.api.CopulaDistribution.rvs.html#statsmodels.distributions.copula.api.CopulaDistribution.rvs )).

Given the current usage of the object API by these libraries, it is relatively straightforward to see how they would adopt an object API if added to the spec.

What is less clear is how they would leverage a functional API. If we are interested in evaluating that option, it would be worth seeing how that API would be used in a downstream library. Should add we probably want to look at something a bit more general than JAX's current API given it is tied to a specific class of PRNGs as pointed out in a few places in this issue already.

**Comment by rkern** [comment]
Another way to think about it is that all PRNG schemes have a certain finite amount of safety margin, and it is a _consumable_ resource. Just serially drawing arrays of numbers from it consumes a small amount of that safety margin. Splits, no matter how implemented, consume _a lot_ of that safety margin.

My general recommendation is to only split when you have to, at the place where you need parallelism. In JAX programs, that's _everywhere_, because JAX is awesome at enabling incredible amounts of parallelism without having to explicitly code it. A splitting-central PRNG scheme fits very well in that environment. The parallelism pays back what you spent in terms of the lowered safety margin. But for most of the other Array API implementations, forcing that amount of splitting is forcing everyone to live with that impaired safety margin for no corresponding benefit.

The issue is not merely the ergonomics of the API. Simply wrapping _a_ stateful API around a splitting-central stateless API does not restore the safety margin built into the more typical stateful PRNG implementations.

**Comment by rgommers** [comment]
> > One other thing to point out: the [JAX docs comparing to NumPy](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#random-numbers) are wildly outdated/unfair, they use the non-recommended (global state) way of using the legacy API.
> 
> I'm happy to update these if you think it would be helpful. In my experience, the global default rng is how most users use numpy's random APIs, despite more recent changes to recommendations in numpy's docs, so it's a useful way to introduce how JAX differs. As for the JAX discussion, whether the seed is global or a mutated `rng` object, I think the bulk of the discussion (regarding side effects, implicit vs. explicit statefulness / pure functions) still holds true.

Thanks @jakevdp. I think it would be nice to either update it, or at least add a note saying that the section talks about legacy APIs and that numpy has had a `default_rng`/`Generator` API that has parallel capabilities and statistical properties & performance similar to (or even better than, xref Robert's analysis above) JAX since 2019. A sentence like _"The Mersenne Twister PRNG is also known to have a [number](https://cs.stackexchange.com/a/53475) of problems, it has a large 2.5Kb state size, which leads to problematic [initialization issues](https://dl.acm.org/citation.cfm?id=1276928). It [fails](http://www.pcg-random.org/pdf/toms-oneill-pcg-family-v1.02.pdf) modern BigCrush tests, and is generally slow."_ looks like it is solely designed to say "see how much better JAX is than NumPy here".

**Comment by rkern** [comment]
I do think it _is_ worth talking about `SeedSequence` spawning there too (even if you do have to address the legacy `np.random.*` stuff as well). It is a place where modern numpy _is_ quite similar to JAX. The recommended ways to achieve parallelism are the basically same in both; the major difference is when and how much parallelism is enforced by the API (and enabled by the rest of JAX).

The other thing I would remove is the discussion about the [sequential-equivalent guarantee](https://jax.readthedocs.io/en/latest/design_notes/prng.html). We have none. If there is a place in our docs that says otherwise, please let me know so I can expunge it. Something like that might have crept into some of the examples, which were mostly written by third-party contributors who might have assumed that the behavior they observed was guaranteed for some reason. Most methods do _happen_ to behave that way because it's generally the easiest thing to implement with a serial PRNG, but we do not guarantee anything about it. For example, if we had a distribution where there was an algorithm that worked well in bulk but had a costly one-time setup and another that worked better for small numbers at a time, we reserve the right to implement a switchover point.

**Comment by jakevdp** [comment]
Hi @rkern - thanks for those comments. Just to be clear, the doc you linked to is not really "documentation" *per se*, but rather a years-old design doc meant to lay-out the motivation for JAX's initial PRNG design (which has evolved since then, and will continue to evolve). For that reason, I don't think I will be updating it, but rather will add a disclaimer at the top mentioning its intent. How does that sound?

**Comment by rkern** [comment]
There might be a couple of other opportunities to standardize _something_, but it's not particularly clear to me what it would actually enable across implementations as different as numpy and JAX, say.

So for example, we could say, informationally, that there are 3 basic flavors of PRNG state flow that an implementation could have, but the standard doesn't specify any one or any of the details about each.

1. Stateful `Generator` with distribution methods (`x = rng.uniform(0, 1)`) (implemented by numpy, dask, maybe others)
2. JAX-like splitting-central pure functional (`key, subkey = split(key); x = uniform(subkey, 0, 1)`, etc.) (implemented by JAX)
3. Copy-and-return-iterated pure functional (`rng, x = uniform(rng, 0, 1)`) (not implemented by anyone that I'm aware of)

But it _could_ standardize on a basic list of methods/functions that an implementation ought to have, along with the semantics of the other arguments not related to the PRNG state flow, and allowing for extra implementation-specific arguments (dask adds `chunks=`, for example). In particular, in the transition from `RandomState.randint()` to `Generator.integers()`, we fixed a number of corner cases that had made it hard to specify the full range of `np.uint64`, for example. It would be a really good idea for all of the implementations to also implement that logic and not suffer from the naive implementation mistakes I made in `randint()`.

This is not likely actually enable one to write significant backend-agnostic code. So the Array API standard might not be the best place for that. Maybe a SPEC is a better instrument? The Array API does have the advantage of having convened the right people.

**Comment by rkern** [comment]
@jakevdp Sure, but it's also in [the tutorial](https://jax.readthedocs.io/en/latest/jax-101/05-random-numbers.html).

**Comment by jakevdp** [comment]
Thanks for pointing that out. To be honest, it's news to me that numpy does not provide a sequential-equivalent guarantee in it's pseudo-random values. I've spent the better part of the last couple decades assuming it did.

**Comment by jakevdp** [comment]
Is there a place in the numpy docs that mentions the lack of a sequential equivalent guarantee in pseudo-random numbers? I'd like to link to it in the discussion. If not, I can link to this thread.

**Comment by edloper** [comment]
On Wed, Jun 8, 2022 at 2:46 PM Jake Vanderplas ***@***.***>
wrote:

> Is there a place in the numpy docs that mentions the lack of a sequential
> equivalent guarantee in pseudo-random numbers? I'd like to link to it in
> the discussion. If not, I can link to this thread.
>
This is the closest I found: https://numpy.org/neps/nep-0019-rng-policy.html

-Edward

**Comment by jakevdp** [comment]
Thanks for the pointers - in that case I think I will link to this thread, because it seems more direct than linking to a long doc and pointing out the omission of the topic in question.

**Comment by rkern** [comment]
@froystig That's all good to hear! Like I said in my introduction, I didn't consider any of those details as immutable black marks against JAX or the overall design, but they did serve as a way to talk about the similarities with `SeedSequence`, the kinds of improvements that are left in the design space (bijections as far as the eye can see!), and what limits are still fundamental to a splitting-centric design (the Birthday Bound). And I'm a long-winded jerk.

**Comment by rkern** [comment]
I don't know all of JAX's rewriting capabilities, but I don't think it can handle whatever it needs to do to make the `Generator`-like API work with the same reordering benefits.

```python
data1 = rng.uniform()
data2 = rng.uniform()
```

What it would have to do here is recognize that `rng` is a `Generator`, so the special case applies, and magically under the covers transform the code to do a couple of splits up at the top before any of the `uniform()` calls.

And even if JAX did have technical capability of doing so, I'm not sure that they want their users to mix mental models like that. It's hard enough to teach people to use one consistently.

**Comment by froystig** [comment]
> they did serve as a way to talk about the similarities with `SeedSequence`, the kinds of improvements that are left in the design space (bijections as far as the eye can see!), and what limits are still fundamental to a splitting-centric design (the Birthday Bound).

@rkern Absolutely – I understood that and it makes sense. I only meant to share some related thoughts and work that's in progress, mostly as an aside. Your comments are super valuable, and this thread is a rare opportunity for us on JAX. Thank you!

**Comment by wangpengmit** [comment]
I'm the author of `tf.random.Generator` and several docs (e.g. [this](https://github.com/tensorflow/community/blob/master/rfcs/20181217-tf2-random-numbers.md) and [this](https://www.tensorflow.org/guide/random_numbers)). TF RNG's current status is using stateless RNGs (similar to JAX's) as the underlying engine, and a mutable counter (implemented as a `tf.Variable`) to advance the RNG state ([code](https://github.com/tensorflow/tensorflow/blob/8508b747acb287898651274d8ee3e800f0a13824/tensorflow/python/ops/stateful_random_ops.py#L543)). `tf.random.Generator` is basically `stateless RNGs + a mutable counter`. It works well in TF because TF is more friendly to mutability than JAX.

Unlike JAX, `tf.random.Generator` supports but doesn't do splitting unless asked to ([code](https://github.com/tensorflow/tensorflow/blob/8508b747acb287898651274d8ee3e800f0a13824/tensorflow/python/ops/stateful_random_ops.py#L913)). It just does counter increment in common (sequential) cases ([code](https://github.com/tensorflow/tensorflow/blob/8508b747acb287898651274d8ee3e800f0a13824/tensorflow/python/ops/stateful_random_ops.py#L575)). I feel splitting is more wasteful of the `(key, counter)` space than counter increment, but I'm not sure.

TF only supports counter-based RNG algorithms and has no plan to support non-counter-based ones.

**Comment by rkern** [comment]
Yes, I can confirm that splitting is more profligate of the state space than the counter increment, especially when you constrain yourself to 64-bit `key`s, if I'm reading the code correctly. I think the same math from my analysis of JAX's PRNG splitting applies (the only real difference is that `tf.random.Generator` will generate the new keys at whatever `counter` it is currently on rather than JAX always starting at `0`, but that doesn't affect the non-invertible mapping argument).

**Comment by ntessore** [comment]
From a user perspective, specifically someone writing libraries which want to support arbitrary array types, I don't see the problem with carrying the random state separately from the random namespace. It is not the slickest interface, but it can be implemented today:

```py
# file: random_numpy.np

def standard_normal(rng, shape=()):
    return rng, rng.standard_normal(size=shape)

def poisson(rng, lam, shape=()):
    return rng, rng.poisson(lam, size=shape)


# file: random_jax.py

from jax import random

def standard_normal(key, shape=()):
    key, subkey = random.split(key)
    return key, random.normal(subkey, shape)

def poisson(key, lam, shape=()):
    key, subkey = random.split(key)
    return key, random.poisson(subkey, lam, shape)


# file: test.py

def random_namespace(rs):
    import sys
    if 'numpy' in sys.modules:
        import numpy as np
        if isinstance(rs, np.random.Generator):
            return __import__('random_numpy')
    if 'jax' in sys.modules:
        import jax
        if isinstance(rs, jax.Array):
            return __import__('random_jax')

def my_function(lam, rs):
    random = random_namespace(rs)
    rs, rv1 = random.standard_normal(rs, shape=(4,))
    rs, rv2 = random.poisson(rs, lam, shape=(4,))
    return rv1 + rv2

import numpy as np
rng = np.random.default_rng()
print(my_function(1.0, rng))
# [ 1.25035394 -0.11511349  1.87203598  2.55088409]

import jax
key = jax.random.PRNGKey(42)
print(my_function(1.0, key))
# [0.43244982 0.8856307  0.06793922 1.4641162 ]
```

For anything more complicated than sampling random variates, particularly anything that requires knowledge of the implementation details of the random state, we accept that we will always have to handle each framework separately.

**Comment by jakevdp** [comment]
The main difference is that `rng` in numpy and `key` in JAX are fundamentally different kinds of objects. In the numpy example, the call to `my_function` mutates `rng` in-place, so that subsequent calls will return different random values. In JAX, functions are pure, and so the call to `my_function` cannot mutate `key`, and subsequent calls with the same key will produce identical values to the first call:
```python
import numpy as np
rng = np.random.default_rng()
print(my_function(1.0, rng))
# [ 1.25035394 -0.11511349  1.87203598  2.55088409]
print(my_function(1.0, rng))
# [3.15990919 1.07582056 1.09202392 0.33987543]

import jax
key = jax.random.PRNGKey(42)
print(my_function(1.0, key))
# [-0.5675502   1.8856307   0.06793922  3.464116  ]
print(my_function(1.0, key))
# [-0.5675502   1.8856307   0.06793922  3.464116  ]
```

**Comment by ntessore** [comment]
Yes, there is probably no universal API for situations where it matters whether or not we are dealing with a stateful RNG or a stateless key.

My point is that it does not preclude us from having a universal (functional) API that can generate random numbers. In my experience, that's also the only situation where you want such an API. For everything more complicated, you need special cases anyway.

**Comment by kgryte** [comment]
Given lack of ecosystem agreement, we are not likely to forge a path forward for standardizing a PRNG API in the array API specification at this time. As such, I will go ahead and close this issue.

We can reopen/revisit if and when we have greater community consensus.

**Comment by ricardoV94** [comment]
> Copy-and-return-iterated pure functional (rng, x = uniform(rng, 0, 1)) (not implemented by anyone that I'm aware of)

This is such an insightful discussion. Just wanted to say that this is actually what we do at PyTensor (descendants of Theano). Our functional API is `next_rng, sample = uniform(rng)`. Then during compilation if a user sets (or we automatically) state that when computing next_rng, rng can be mutated, we simply do what numpy does. By default we copy the rng before using (wasteful of course, but used for debugging, doing intermediate evaluations without affecting the final graph).
